lib/publiccloud/basetest.pm

```diff
@@ -148,49 +173,56 @@ sub finalize {
+          or record_info('FAILED upload_logs', $text . "Failed _upload_logs (ref.: \$self->_upload_logs) --\n $@");
     }
-    diag('Public Cloud finalize: 2nd check passed.');
-
-    eval { $self->_upload_logs(); } or record_info('FAILED', "\$self->_upload_logs() failed -- $@", result => 'fail');
-
     # We need $self->{run_args} and $self->{run_args}->{my_provider}
-    if ($self->{run_args} && $self->{run_args}->{my_provider}) {
-        diag('Public Cloud finalize: Ready for provider teardown.');
+    if ($self->{run_args}->{my_provider}) {
+        diag($text . 'Ready for provider teardown.');
         # Call the provider teardown
-        eval { $self->{run_args}->{my_provider}->upload_boot_diagnostics() } or record_info('FAILED', "\$self->{run_args}->{my_provider}->upload_boot_diagnostics() failed -- $@");
-        eval { $self->{run_args}->{my_provider}->teardown() } or record_info('FAILED', "\$self->{run_args}->{my_provider}::teardown() failed -- $@", result => 'fail');
-        diag('Public Cloud finalize: The provider teardown finished.');
+        eval { $self->{run_args}->{my_provider}->upload_boot_diagnostics() }
+          or record_info('FAILED upload_boot_diagnostics', $text . "Failed provider upload_boot_diagnostics (ref.: \$self->{run_args}->{my_provider}->upload_boot_diagnostics) --\n $@");
+        eval { $self->{run_args}->{my_provider}->teardown() }
+          or record_info('FAILED teardown', $text . "Failed provider teardown (ref.: \$self->{run_args}->{my_provider}::teardown) --\n $@");
+        diag($text . 'The provider teardown finished.');
     } else {
-        diag('Public Cloud finalize: Not ready for provider teardown.');
+        diag($text . 'Not ready for provider teardown.');
+        return 0;
     }
+    return 1;
 }
 
 sub _upload_logs {
     my ($self) = @_;
     my $ssh_sut_log = '/var/tmp/ssh_sut.log';
-
-    diag('Public Cloud _upload_logs: $self->{run_args}=' . $self->{run_args}) if ($self->{run_args});
-    diag('Public Cloud _upload_logs: $self->{run_args}->{my_instance}=' . $self->{run_args}->{my_instance}) if ($self->{run_args}->{my_instance});
-    unless ($self->{run_args} && $self->{run_args}->{my_instance}) {
-        die('Public Cloud _upload_logs: Either $self->{run_args} or $self->{run_args}->{my_instance} is not available. Maybe the test died before the instance has been created?');
-    }
-
-    script_run("sudo chmod a+r " . $ssh_sut_log);
-    upload_logs($ssh_sut_log, failok => 1, log_name => $ssh_sut_log . ".txt");
-
-    my @instance_logs = ('/var/log/cloudregister', '/etc/hosts', '/var/log/zypper.log', '/etc/zypp/credentials.d/SCCcredentials');
-    for my $instance_log (@instance_logs) {
-        $self->{run_args}->{my_instance}->ssh_script_run("sudo chmod a+r " . $instance_log, timeout => 0, quiet => 1);
-        $self->{run_args}->{my_instance}->upload_log($instance_log, failok => 1, log_name => $instance_log . ".txt");
+    return 0 unless ($self->{run_args});
```

Comment:
before that we were uploading ssh log even when run_args are not defined , now we will not do this anymore . can you elaborate why you think this is necessary ?

Reply:
This github difference matching here is really unclear, I didn't realy removed the bove red-part, that is reapplied by the git parser here below as an addition.
Here I exit the routine when run_args is empty, because no other depending object could be present.

Reply:
Yes and this was at the origin of this p.oo ticket, see [19057129](https://openqa.suse.de/tests/19057129#step/prepare_instance/146), not useful at all its **stopping `die`** that over-wrote the original fault that in turn triggered in post-fail-hook `finalize` and this inner `_upload_logs` subroutine.

Here inside, only `$self->{run_args}->{my_instance}` and its parent `$self->{run_args}` are managed, Therefore:
- if the parent runargs is empty, like here, no sub-childen can run at all in the routine, then return;
- if my_instance is empty, the same shall return, never stopping the test for that. 

Just noticing now, that one only log `$ssh_sut_log` is indipendent, and I will change the code to execute it on top:

Reply:
New version is coming soon.